drivers/flash/flash_shell.c

```diff
@@ -153,12 +156,14 @@ static int cmd_read(const struct shell *shell, size_t argc, char *argv[])
 		cnt = 1;
 	}
 
-	while (cnt--) {
-		u32_t data;
+	cnt *= sizeof(u32_t);
+	for (upto = 0; upto < cnt; upto += todo) {
+		u8_t data[SHELL_HEXDUMP_BYTES_IN_LINE];
 
-		flash_read(flash_dev, addr, &data, sizeof(data));
-		shell_print(shell, "0x%08x ", data);
-		addr += sizeof(data);
+		todo = MIN(cnt, SHELL_HEXDUMP_BYTES_IN_LINE);
```

Comment:
i think it should be `cnt - upto`

Reply:
oops missed this. Fixed now.

What is the best way to write a test for this? I could use the dummy shell backend perhaps?

Reply:
not sure how flash works on qemu target and if that can be used for testing. @nvlsianpu ?

Reply:
Eek I don't want to use qemu. I just wondered if there was a native/posix driver.

Reply:
It does work on qemu an posix using flash simulator driver.

Reply:
Qemu is a large overhead for testing a simple thing. It looks like I can use posix for this, although I do see what are apparently bugs in the driver.

https://github.com/zephyrproject-rtos/zephyr/issues/25109

Reply:
I found a way to do a native test